src/dotnet-roslyn-tools/DartTest/DartTest.cs

```diff
@@ -0,0 +1,237 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Net.Http.Json;
+using System.Text;
+using System.Text.Json.Nodes;
+using System.Threading.Tasks;
+using LibGit2Sharp;
+using Microsoft.Azure.Pipelines.WebApi;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.RoslynTools.Products;
+using Microsoft.RoslynTools.Utilities;
+using Microsoft.RoslynTools.VS;
+using Microsoft.TeamFoundation.SourceControl.WebApi;
+using static System.Net.WebRequestMethods;
+
+namespace Microsoft.RoslynTools.DartTest;
+
+internal static class DartTest
+{
+    public static async Task<int> RunDartPipeline(
+        string productName,
+        RemoteConnections remoteConnections,
+        ILogger logger,
+        int prNumber,
+        string? sha)
+    {
+        var cancellationToken = CancellationToken.None;
+        var azureBranchName = $"dart-test/{prNumber}";
+        var product = VSBranchInfo.AllProducts.Single(p => p.Name.Equals(productName, StringComparison.OrdinalIgnoreCase));
+
+        // If the user doesn't pass the SHA, retrieve the most recent from the PR.
+        if (sha is null)
+        {
+            sha = await GetLatestShaFromPullRequest(product, remoteConnections.GitHubClient, prNumber, logger, cancellationToken).ConfigureAwait(false);
+            if (sha is null)
+            {
+                logger.LogError("Could not find a SHA for the given PR number.");
+                return -1;
+            }
+            else
+            {
+                logger.LogError("A SHA is required for the given product.");
+                return -1;
+            }
+        }
+
+        var directory = await PushPRToInternal(product, prNumber, azureBranchName, logger, sha, cancellationToken).ConfigureAwait(false);
+        var repositoryParams = new Dictionary<string, RepositoryResourceParameters>
+            {
+                {
+                    "self", new RepositoryResourceParameters
+                    {
+                        RefName = $"refs/heads/{azureBranchName}",
+                        Version = sha
+                    }
+                }
+            };
+
+        var runPipelineParameters = new RunPipelineParameters
+        {
+            Resources = new RunResourcesParameters
+            {
+
+            },
+            TemplateParameters = new Dictionary<string, string> { { "prNumber", prNumber.ToString() }, { "sha", sha } }
+        };
+
+        await remoteConnections.DevDivConnection.TryRunPipelineAsync(product.DartLabPipelineName, repositoryParams, runPipelineParameters, logger).ConfigureAwait(false);
+        CleanupDirectory(directory, logger);
+        return 0;
+    }
+
+    private static async Task<string?> GetLatestShaFromPullRequest(IProduct product, HttpClient gitHubClient, int prNumber, ILogger logger, CancellationToken cancellationToken)
```

Comment:
This will likely be different for repos hosted in ADO. We could do a follow up which fully enabled this for ADO. Would push logic into the appropriate `IHost`.